ksql-rest-app/src/main/java/io/confluent/ksql/rest/util/ClusterTerminator.java

```diff
@@ -0,0 +1,161 @@
+            ksqlEngine.terminateQuery(persistentQueryMetadata.getQueryId(), true);
+          }  else {
+            queryMetadata.close();
+          }
+        }
+    );
+
+    // if we have the explicit list of stream/table to delete.
+    final MetaStore metaStore = ksqlEngine.getMetaStore();
+    if (sourcesList.isEmpty()) {
+      metaStore.getAllStructuredDataSources().forEach((s, structuredDataSource) ->
+          deleteSource(s, structuredDataSource, ksqlEngine));
+    } else {
+      if (isKeep) {
+        metaStore.getAllStructuredDataSources().forEach(
+            (sourceName, structuredDataSource) -> {
+              if (!sourcesList.contains(sourceName)) {
+                deleteSource(sourceName, structuredDataSource, ksqlEngine);
+              }
+            }
+        );
+      } else {
+        sourcesList
+            .forEach(sourceName -> deleteSource(
+                sourceName,
+                metaStore.getSource(sourceName),
+                ksqlEngine));
+      }
+    }
+  }
+
+  // This is needed because the checkstyle complains if we create this in place.
+  private List<QueryMetadata> getListForSet(
+      final KsqlEngine ksqlEngine
+  ) {
+    final List<QueryMetadata> queryMetadataList = new ArrayList<>();
+    queryMetadataList.addAll(ksqlEngine.getAllLiveQueries());
+    return queryMetadataList;
+  }
+
+  private void deleteSource(
+      final String sourceName,
+      final StructuredDataSource structuredDataSource,
+      final KsqlEngine ksqlEngine) {
+    final DdlCommand ddlCommand = new DropSourceCommand(
+        getAbstractStreamDropStatement(sourceName,
+            structuredDataSource.getDataSourceType() == DataSourceType.KSTREAM),
+        structuredDataSource.getDataSourceType(),
+        ksqlEngine.getTopicClient(),
+        ksqlEngine.getSchemaRegistryClient(),
+        true
+    );
+    final DdlCommandExec ddlCommandExec = ksqlEngine.getDdlCommandExec();
+    ddlCommandExec.execute(ddlCommand, false);
+  }
+
+  private AbstractStreamDropStatement getAbstractStreamDropStatement(
+      final String sourceName,
+      final boolean isStream) {
+    if (isStream) {
+      return new DropStream(QualifiedName.of(sourceName), false, true);
+    } else {
+      return new DropTable(QualifiedName.of(sourceName), false, true);
+    }
+  }
+
+  private void deleteCommandTopic(
+      final KsqlConfig ksqlConfig,
+      final KsqlEngine ksqlEngine,
+      final Consumer<CommandId, Command> commandConsumer,
+      final Producer<CommandId, Command> commandProducer) {
+    // Delete the command topic
+    commandConsumer.close();
+    commandProducer.close();
+    final String ksqlServiceId = ksqlConfig.getString(KsqlConfig.KSQL_SERVICE_ID_CONFIG);
+    final String commandTopic = KsqlRestConfig.getCommandTopic(ksqlServiceId);
+    final KafkaTopicClient kafkaTopicClient = ksqlEngine.getTopicClient();
+    try {
+      ExecutorUtil.executeWithRetries(
+          () -> kafkaTopicClient.deleteTopics(
```

Comment:
What happens to the other servers/consumers when we do this? Do they still consume the topic till the end?

Reply:
No, the topic is being deleted. It's gone. Poof!

Reply:
Right but what happens to those consumers when they call ```poll```? They get an exception?

Reply:
True. Maybe we shouldn't both writing TERMINATE to the command topic. Instead, just use the deletion of the command topic as the indication the other nodes should exit.

Reply:
Assuming the deletion of the command topic as termination sign is a good option but we won't be able to use `KEEP` or `DELETE` lists as described here.

Reply:
One option would be to consider the cluster termination process as the following steps:
1. shutdown all of the servers except one.
2. submit the `TERMINATE CLUSTER` command and have only the last server to do the clean up.
3. When the command topic is deleted shutdown the last server.

@rodesai @big-andy-coates what do you think?

Reply:
Only one node needs to handle the `KEEP` / `DELETE` right? Doesn't make sense for them all to try and delete the same topics.  There I still think deleting the command topic is the cleanest way of signalling to the other nodes that they should shut down.  Shutdown sequence should be:

1. Delete the command topic, (which initiates other nodes to shutdown). If the command topic has already been deleted, just shutdown, (as it means multiple servers have received a TERMINATE call and hence we've a race condition - so just log it and shutdown).
1. Maybe a configurable grace period. (Either in TERMINATE payload and/or via server config - this will reduce noise in the logs of the other servers from having other topics pulled out from underneath them.
1. The single server that received TERMINATE and deleted the command topic should then honour the KEEP/DELETE list.